src/Build/BackEnd/Components/SdkResolution/NuGet/NuGetSdkResolverBase.cs

```diff
@@ -0,0 +1,98 @@
+        /// </summary>
+        public const string NuGetAssemblyPathEnvironmentVariableName = "MSBUILD_NUGET_PATH";
+
+        /// <summary>
+        /// Attempts to locate the NuGet assemblies based on the current <see cref="BuildEnvironmentMode"/>.
+        /// </summary>
+        private static readonly Lazy<string> NuGetAssemblyPathLazy = new Lazy<string>(() =>
+        {
+            // The environment variable overrides everything
+            string basePath = Environment.GetEnvironmentVariable(NuGetAssemblyPathEnvironmentVariableName);
+
+            if (!String.IsNullOrWhiteSpace(basePath) && Directory.Exists(basePath))
+            {
+                return basePath;
+            }
+
+            if (BuildEnvironmentHelper.Instance.Mode == BuildEnvironmentMode.VisualStudio)
+            {
+                // Return the path to NuGet under the Visual Studio installation
+                return Path.Combine(BuildEnvironmentHelper.Instance.VisualStudioInstallRootDirectory, PathToNuGetUnderVisualStudioRoot);
+            }
+
+            // Expect the NuGet assemblies to be next to MSBuild.exe, which is the case when running .NET CLI
+            return BuildEnvironmentHelper.Instance.CurrentMSBuildToolsDirectory;
+        });
+
+        static NuGetSdkResolverBase()
+        {
+#if FEATURE_APPDOMAIN
+            AppDomain.CurrentDomain.AssemblyResolve += AssemblyResolve;
+#else
+            AssemblyLoadContext.Default.Resolving += AssemblyResolve;
+#endif
+        }
+
+        /// <summary>
+        /// A custom assembly resolver used to locate NuGet dependencies.  It is very important that we do not ship with
+        /// these dependencies because we need to load whatever version of NuGet is currently installed.  If we loaded our
+        /// own NuGet assemblies, it would break NuGet functionality like Restore and Pack.
+        /// </summary>
+        private static Assembly AssemblyResolve(
+#if FEATURE_APPDOMAIN
+            object sender,
+            ResolveEventArgs args)
+        {
+            AssemblyName assemblyName = new AssemblyName(args.Name);
+#else
+            AssemblyLoadContext assemblyLoadContext,
+            AssemblyName assemblyName)
+        {
+#endif
+            // Only load NuGet.* and Newtonsoft.Json assemblies if found
+            if (assemblyName.Name.StartsWith("NuGet.", StringComparison.OrdinalIgnoreCase) || assemblyName.Name.Equals("Newtonsoft.Json", StringComparison.OrdinalIgnoreCase))
+            {
+                string assemblyPath = Path.Combine(NuGetAssemblyPathLazy.Value, $"{assemblyName.Name}.dll");
+
+                if (File.Exists(assemblyPath))
+                {
+#if !FEATURE_APPDOMAIN
+                    return assemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+#elif !CLR2COMPATIBILITY
+                    return Assembly.UnsafeLoadFrom(assemblyPath);
+#else
+                    return Assembly.LoadFrom(assemblyPath);
```

Comment:
How do these loads work in the face of deployed NuGet DLLs that are a different version from what we built against? Is it different between Core and full framework?

Reply:
They are effectively assembly binding redirects.  If we compile against 4.3 we can safely load 4.5 instead.  As long as there are no breaking API changes, it should work.